[RFC] cleanup some deprecated options

Dudemanguy opened the pull request:
Felt like doing some janitor work again. All of these were from mpv 0.37 or earlier, and I see no reason to continue leaving these deprecated/useless options in the codebase for years. Let me know if you think if any of these are too soon. I also didn't document the removal of the `OPT_REMOVED` ones because mpv fails opening if you use any one of those so it's not really a terribly relevant interface change imo. Just the removal message goes away.

Didn't touch the `--gamma-auto` or `--gamma-factor` stuff since we already have a PR for that (#14369).

Dudemanguy commented:
I would definitely insist on removing at least `--fps`, `--cache-dir`, and `--sub-forced-only` given that the names of those options are so bad that it's the opposite of user friendly.

Dudemanguy commented:
I personally think `OPT_REMOVED` is only useful in very specific situations since it still treats the option as invalid but just gives an additional message. Usually `OPT_REPLACED` or just adding a `.deprecation_message` will suffice. It's useful for when the semantics of some option completely changes and you want to inform users how to get back the old behavior. The recent `--focus-on-open` is a good example of that. But as time goes on, nobody needs to know that `--gamut-warning` used to be a separate option and it's just dead code that should be removed.

Dudemanguy commented:
>because these aliases suddenly stopped working without any information about replacements.

What? Deprecated aliases inform users about replacements. Some of them in that patch series were spitting out warnings for several **years**. If someone doesn't bother to spend two seconds to update/change the thing that was removed, why is this our problem?

>If you want to continue this practice (which I personally find no practical benefit)

If nobody actually ever removes these things, then there's no point in ever deprecating them in the first place.

kasper93 commented:
> What? Deprecated aliases inform users about replacements. Some of them in that patch series were spitting out warnings for several years. If someone doesn't bother to spend two seconds to update/change the thing that was removed, why is this our problem?

Please consider that average user does not update mpv every week. In some cases the users migrate multiple version forward. Like when updating Ubuntu LTS, you jump form 0.34.1 to 0.37 and if you for some reason skipped previous LTS, you start from 0.32.

Also consider there are guides / configs laying around in the forums that are outdated and by not removing those aliases we simply allow users to easily port those configs forward, for better or for worse.

> What? Deprecated aliases inform users about replacements. Some of them in that patch series were spitting out warnings for several years. If someone doesn't bother to spend two seconds to update/change the thing that was removed, why is this our problem?

Please consider that the average user does not update mpv every week. In some cases, users migrate multiple versions forward. For example, when updating Ubuntu LTS, you jump from 0.34.1 to 0.37, and if you skipped the previous LTS, you start from 0.32.

Also, consider that there are guides and configs lying around in forums that are outdated. By not removing those aliases, we simply allow users to easily port those configs forward, for better or for worse.
> If nobody actually ever removes these things, then there's no point in ever deprecating them in the first place.

I agree with @na-na-hi on this. Neither `OPT_REMOVED` nor `OPT_REPLACED` are intended to be removed. Their exact purpose is to mark old options for future users. `OPT_REMOVED` is after deprecated period, so depreciated things already were removed.

There are two types of option removal:

1. The option (and the associated feature) gets deprecated, the `deprecation_message` is set, and after a few releases, the option is finally removed and marked with `OPT_REMOVED`, possibly with a useful message or simply to notify that the option was removed but once existed.
2. The option is renamed, for whatever reason, but the functionality of this option is not changed. `OPT_REPLACED` is used to map it to the new version with a nagging depreciation warning for the user. A replaced option like that may be transformed into `OPT_REMOVED` if changes are made to the new option, as it shouldn't affect the old alias.

In both of these cases, `OPT_REMOVED` and `OPT_REPLACED` are the end. They should stay indefinitely in the code. All related code is already removed, and only the documentation in the form of these macros remains. There is no longer any deprecated code to remove. Only the nagging message for the user remains, but this is a good thing.

Dudemanguy commented:
>OPT_REPLACED options were never meant to be removed
>Neither OPT_REMOVED nor OPT_REPLACED are intended to be removed.

I'm sorry but this is frankly ridiculous. The default message when you use `OPT_REPLACED` literally has a big, fat "might be removed in the future".

kasper93 commented:
> I'm sorry but this is frankly ridiculous. The default message when you use OPT_REPLACED literally has a big, fat "might be removed in the future".

Sure it can be removed after some time and marked with `OPT_REMOVED`.

kasper93 commented:
> Why is this suddenly a rule now? In https://github.com/mpv-player/mpv/pull/12436, I removed a bunch of crap including OPT_REPLACED and no one complained.

The fact no one complained doesn't make this everyone agreed.

> I'd totally get it if you guys said something like "option X is still used quite a bit in my opinion, let's hold off for a bit", but it basically sounds like a highly unreasonable amount of backwards compatibility which is not something mpv ever did. Especially not for dumb obscure options no one uses.

In fact what I'm saying mostly is that `OPT_REMOVED` shouldn't be removed ever. When `OPT_REPLACED` is removed is different story, but by "removed" I mean replaced with `OPT_REMOVED`.

In wm4's words:
> OPT_REMOVED can be used to inform the user of alternatives or reasons for the removal, which is better than an option not found error.

It is already removed, there is not much more we can do. And showing a meaningful message instead of `opt not found`  is good thing and I don't consider it backward compatible, just something we get for free. (technically one line of code, but it is still free)

Dudemanguy commented:
>The fact no one complained doesn't make this everyone agreed.

I mean your own thumbs up is currently there so I presume you agreed. This PR is not really any different in principle. The options here are just not nearly as old as the ones there.

As for `OPT_REMOVED` like I said earlier, it's only really useful in my opinion for niche situations where option semantics change in a way that cannot be simply mapped to the new option name. If you want to simply remove something completely with no real replacement, `deprecation_message` and making the option do nothing is a strictly better choice. Of course, there is no actual harm in leaving these in the codebase for a bit when something was recently moved, but eventually it becomes useless information. e.g. ``--vulkan-disable-events`` was removed in 7f5541fc3c440a4aa5fdec2899c02d1caf7fb227. The message that is currently printed is "Unused". Is this really great, wonderful information that mpv needs to print until the end of time?

kasper93 commented:
> I mean your own thumbs up is currently there so I presume you agreed.

Like I said, na-na-hi made a good point. I don't think it is strictly bad to remove those options, but now I see that we can do better without any additional cost in practice.

> If you want to simply remove something completely with no real replacement, deprecation_message and making the option do nothing is a strictly better choice.

I feel like it would be inconsistent with the mpv cli options. Unknown options return error, removed should too. The only gain of `OPT_REMOVED` is additional message that might be useful for user.

No one reads warnings, until they are errors.

> The message that is currently printed is "Unused". Is this really great, wonderful information that mpv needs to print until the end of time?

Of course not, I feel like it could be resolved easily with OPT_REMOVE_IN(0.35) which could print something like "removed in mpv v0.35" which is infinitely more useful and actually can be left until the end of time.

Dudemanguy commented:
>I feel like it would be inconsistent with the mpv cli options. Unknown options return error, removed should too. The only gain of OPT_REMOVED is additional message that might be useful for user.

So what are you suggesting that instead of using .deprecation_message when removing an option we should just hard error when we remove something (see `--drm-atomic` for example). So much for all the effort on keeping a deprecation period. Or do you think we should do some extra long transition from a `.deprecation_message` -> `OPT_REMOVED` -> delete it for real? I find the middle step in this case to be completely pointless and redundant.

>No one reads warnings, until they are errors.

If a user ignores a big fat warning for an extended period of time, I fail to see why this is our problem. It's doubtful such a person would even bother to read an error.

>I feel like it could be resolved easily with OPT_REMOVE_IN(0.35) which could print something like "removed in mpv v0.35" which is infinitely more useful and actually can be left until the end of time

Why is this useful? Does anybody really need to know that mpv used to have some [insert random debugging option here] 20 versions later in the future?

>This is only there to discourage users from using that deprecated alias.

Yes and eventually the endgoal is to make them unable to use it altogether. That's what deprecation means. Look if you really think `OPT_REPLACED` should never ever be removed, then the default message should be completely changed and such options shouldn't be considered deprecated.

>This is a small subset of cases (options now does nothing and has no replacement) compared to all of the other useful information removed.

Useful? Why does any user need to know that `--gamut-clipping` used be a thing? What are we catering to some weirdo person that decides to upgrade from debian oldstable to master in one go? If some user is using some outdated meme config on the internet, we're only doing them a favor by breaking it.

>In summary, is there any benefit with removing them? 

Yes. As stated earlier, I consider the existing deprecated alias `--fps`, `--cache-dir`, `--cache-unlink-files`, and `--sub-forced-only` all to be the opposite of user friendly (user hostile if you will). These options all have terrible names and make the codebase actively worse by existing. In particular, I've *never* seen a user in the wild use `--sub-forced-only` correctly. I don't know if `--sub-forced-events-only` is the greatest name ever but at least I haven't seen anyone misuse it (or rather use it at all given it's super niche) yet. Removing it is for their own good. There is **ample** reason to remove the aliases for at least the aforementioned options given the deprecation has been more than long enough (at least 2 versions).

Nobody uses `--demuxer-cue-codepage`. I am highly confident because `demuxer_cue` was completely broken for an entire version and nobody ever noticed or complained. I only found it by accident and fixed it later. The subset of people that actually would read cuesheets with mpv and pass specific codepages is probably 0. I'm singling out this option in particular because it additionally causes pointless additional bloat by having an extra `m_sub_options` struct for the sole purpose of printing a warning that nobody will see. There's several other options in here that are also useless. Nobody uses `--cdda-toc-bias`. Nobody uses `--drm-atomic`. I highly doubt anyone uses `--vo-sixel-exit-clear`.

All in all, I find this entire discussion completely ludicrous and it's clear we're not going to see eye to eye on this. If you were just telling me "hey dudemanguy, you're removing thing X too fast, it should stick around longer", then we can have a discussion. But instead apparently mpv should be changing it's philosophy to keep atrocious deprecated options like `--fps` in the codebase forever. Or completely useless bloat options that no one uses because we might hypothetically break some rando on the internet that copied some meme config from 10 years ago. mpv has never worked this way. Sure I did a big cleanup recently, but before that plenty of options were commonly removed in releases. Just scroll through the release history a bit.

It doesn't matter to me if having `OPT_REMOVED`  or `OPT_REPLACED` is "not a maintenance burden". If the code is not useful and outlived its purpose, it should be removed.

kasper93 commented:
> So what are you suggesting that instead of using .deprecation_message when removing an option we should just hard error when we remove something (see --drm-atomic for example).

I think there is misunderstanding. `--drm-atomic` was deprecated before, now (this pr) it is removed. You deprecate first, than you remove.

> Or do you think we should do some extra long transition from a .deprecation_message -> OPT_REMOVED -> delete it for real?

No it should be normal deprecation period. `OPT_REMOVED` is there to notify the user afterwards. That's exactly the point, we can have faster removal, even of highly used options, by enforcing `OPT_REMOVED` that clearly guide user what to do. There is no guessing if the option is used by users or not.

> Look if you really think OPT_REPLACED should never ever be removed, then the default message should be completely changed and such options shouldn't be considered deprecated.

User messages doesn't have to reflect perfectly the state of development. User messages should be understandable and clear for users. You don't explain rocket science to 5yo, you only say that rockets work.

> Nobody uses...

Assumptions... last time we made an assumption about `loadfile` usage it got little bit of backlash. It is not like we have any statistic how mpv/libmpv is used. Also I would unify depreciation process to be uniform for all options/properties and not based of assumptions.

> Why is this useful? Does anybody really need to know that mpv used to have some [insert random debugging option here] 20 versions later in the future?
> [...]
> break some rando on the internet that copied some meme config from 10 years ago. mpv has never worked this way.
> [...]
> from debian oldstable to master in one go?

Sorry, but I think you tunnel vision a bit. If we are strict about 2 releases policy for deprecation things removal, it is perfectly possibly that Billy, who doesn't know much, would have theirs config broken on Ubuntu LTS to LTS update.

In such cases user can skip deprecation period, which is fine, we cannot sync to release cycles of other distorts. But here comes `OPT_REMOVED` with helpful message, that option X was removed and possibly alternative.

If you think it is not good to annotate removed option, please remove the `OPT_REMOVED` macro and remove options completely after depreciation period.

Note that @na-na-hi suggest to keep `OPT_REPLACED` indefinitely as it is helpful for users and doesn't add any maintenance cost. On this topic I can agree both ways, if you want to remove options after depreciation period, to force users to tinker with their configs, so be it. All I'm saying that there is still value in keeping them as `OPT_REMOVED`. Note that removing both would require user to not only update their config, but also search through commit histrory / interface changes to understand why the option is no longer working.

> it's clear we're not going to see eye to eye on this.

We don't have to, but it is good to have all the arguments laid down.

Dudemanguy commented:
>`OPT_REMOVED` is there to notify the user afterwards.

Again that's totally fine in some cases but as I've repeatedly said, there's a point where this is useless info that nobody needs. e.g. no user today is going to somehow type `--xineramascreen` and get confused because we removed the message about it getting replaced by `--screen`. It's a temporary guide that becomes obsolete with time.

>User messages doesn't have to reflect perfectly the state of development. User messages should be understandable and clear for users.

"Warning: option --X was replaced with --Y and might be removed in the future." sounds 100% clear and understandable to me. If our policy is actually to never remove `--Y`, then the message should be changed to reflect that.

>Assumptions... last time we made an assumption about loadfile usage it got little bit of backlash.

Sure, I made a miscall on `loadfile`. The command was certainly commonly used but I did not expect the last argument to be used much so the breaking change for a more, imo, sensible argument order was worth it to me. Maybe I shouldn't have done that but oh well. But again, we removed literally dozens of options and deprecations in #12436 with virtually no complaints.

>Also I would unify depreciation process to be uniform for all options/properties and not based of assumptions.

I don't agree. The importance of options/properties is not uniform thus the deprecation process cannot be uniform. It is inherently a sliding scale of what is considered import and what is not. Well unless you want to use an ungodly long deprecation period for everything I suppose. Let's say that for some bizarre reason we decided to completely break `--slang`. Yes it's an assumption, but I think it is safe to say this is one of the most commonly used mpv options. After all, there was that one time we broke its behavior on accident and there was nonstop issues and negative feedback. Anyways whatever deprecation period we use in this case would have be massive. 

On the other hand, let's take `--cdda-toc-bias` again. Why am I pretty sure no one uses this? Because it would have fixed #8777 and probably #1614 and no one ever noticed. I didn't either until I actually dove into the code and discovered it's stupid that this is an option in the first place and should just always be done. I probably could have immediately removed the option and no one would have cared but fine let's be nice and have a brief deprecation period which has been more than long enough in my view.

>If we are strict about 2 releases policy for deprecation things removal, it is perfectly possibly that Billy, who doesn't know much, would have theirs config broken on Ubuntu LTS to LTS update.

This is just an argument for possibly a longer deprecation period for something which is totally fine. We can go on a case-by-case if anyone thinks it is important enough.

>If you think it is not good to annotate removed option, please remove the OPT_REMOVED macro and remove options completely after depreciation period.

As stated before, `OPT_REMOVED` still has a use when the semantics of an option changes and is an immediate breaking change. That happens sometimes. For something that has been deprecated for a long enough time and has been warning users, no I don't think using the `OPT_REMOVED` macro after removing the option completely is worth it. If there is a concern that users may not be aware of the option change for whatever reason, then the answer is simply to extend the deprecation period.

Dudemanguy commented:
>I'm sure you know what the word "might" means.

It also obviously doesn't mean "never". Don't be stupid.

>What you're doing here is exactly the opposite of "doing them a favor" by removing the message which informs the user about the replacement option to use, so that he can fix his config.

`gamut-clipping` was removed several versions ago and replaced by a completely new option with different semantics. Nobody is going to organically type `--gamut-clipping` today. The only way this would happen is because someone is blindly copying some config they found on the internet. Said person should be forced to read the manual and not blindly copy garbage they don't understand.

>That's because the documentation of that option was very unclear. See #13804 (comment). This is the real reason why that option was misused. It has little to do with the naming of the option.

Man now this is an amazing bad faith argument. Don't be ridiculous. `--sub-forced-only` quite obviously sounds like "only select forced subtitles" even without looking at the manual. Both the option name and the manual entry were atrocious. Come on at least give me that. You didn't even attempt to address the other ones I named either or do you actually think `--cache-dir` is a good name?

>Lots of OPT_REPLACED entries you removed in #12436 had been "deprecated" for more than 10 years.

It's not a question. They **were** deprecated. 10 years is **way** more than enough time to remove them. I guess you think that `OPT_REPLACED` should apparently last forever. Again, open up a PR and change the wording of the deprecation message so that bad developers like me don't try to remove it. If you're bored, you can go through the entire mpv commit history and see if I'm the only person who ever removed an `OPT_REPLACED` deprecation. Maybe so, and then you can get me on this technicality. But I would say "spiritually", it's in line with normal development practices. Any `OPT_REPLACED` option is a deprecation. Deprecations are subject to removal after enough period of time. Everyone does this including mpv (sometimes in a very user-unfriendly way too).

>It was ADDED in 7fb5df0 specifically because it's "more userfriendly."

Nobody is contesting that the functionality can be useful. The only thing that is being argued here is that such options have had a sufficient deprecation period and the mapping has outlived its practical use.

>I have explained again and again that removing useless options doesn't preclude providing OPT_REMOVED entries.

And I have explained again and again that these options in question have already been providing deprecation messages for a long time.

>So no practical benefit.

Yes if you conveniently ignore the vast majority of my response to that question and give some wild bad faith arguments, sure.

>So what's the priority here? Removing useless code and inform users immediately with OPT_REMOVED, or extending the deprecation period and keep the useless bloat around? 

Why is this so hard to understand? I've explained what I think `OPT_REMOVED` is good for like 5 different times by now and I'm not going to type it again. Scroll up.

srk24 commented:
Thank you for the work of every developer. I'd like to share some thoughts as an ordinary user. Take the example of --cache-dir mentioned earlier. According to the OP's suggestion, it would be deleted directly, and users would need to modify it themselves. However, when I checked the wiki, I couldn't find any guidance on how to migrate. (In fact, with enough information, I could have known that I should set its sub-option, such as --icc-cache-dir. )

hooke007 commented:
Check https://github.com/mpv-player/mpv/blob/master/DOCS/interface-changes.rst

srk24 commented:
> Check https://github.com/mpv-player/mpv/blob/master/DOCS/interface-changes.rst ![image](https://private-user-images.githubusercontent.com/41094733/341317173-66db79f4-6a64-4231-bf28-0b3b9001b852.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4Njk0MzMsIm5iZiI6MTcxODg2OTEzMywicGF0aCI6Ii80MTA5NDczMy8zNDEzMTcxNzMtNjZkYjc5ZjQtNmE2NC00MjMxLWJmMjgtMGIzYjkwMDFiODUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjIwVDA3Mzg1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkYmEyMTNhMmFlOWNiZGIxZjU1ODdhNzlkMGQyNTYyYzQxYzA5OGZhNzg3Y2ZkNTIyMzM0NjYwNGQyM2RlZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NEwM14QmCZgXKZKNqiQ_TmqKou4iFH-l8ikNmkh3_qQ)

Thx! I get it. I forgot to read this docs, always search in [manual](https://mpv.io/manual/master/), my fault

avih commented:
Without getting into any discussion, I'd just share my point of view:

Keeping OPT_REMOVED for all removed options is useful to the user and doesn't have any maintenance burden, therefore I support keeping OPT_REMOVED forever for all removed options, except for reasons I can't forsee right now which neccessitate deleting it completelty.

That's not deprecation, because it's already past the deprecation preiod and already removed as a functional option, therefore it only served as a note to users "this doesn't exist anymore, do this or that instead", which is cheap for us and useful to the users.

kasper93 commented:
> I don't agree. The importance of options/properties is not uniform thus the deprecation process cannot be uniform. It is inherently a sliding scale of what is considered important and what is not.

Why? I understand what you are saying, but why do you think longer deprecation time solves anything? We have already pretty long period, that certainly gives more than enough time for anyone to adjust to new changes if they want to.

Deprecation period is to allow users to migrate, if they want. And by having uniform policy it would be clear when/how the option is removed/replaced.

Even if we were to "break" `--slang`, two full releases is enough notification period for this. It may be one of the most used options. But having it linger for years, would not make it more convenient. Changes/Removal is mostly for developers to fix/clean the code, depreciation period is for users to adjust to those changes, shorter or no depreciation period means new mpv version are adopted slower as the breaking changes have to be resolved first.

> On the other hand, let's take --cdda-toc-bias again. Why am I pretty sure no one uses this?

And why does it matter? Sure if it is not possible to have deprecation period. But otherwise why make specific exceptions of the rules, just because we assume it is not used? What does it help with, why cannot it go through normal period? To make us feel better that it was removed quicker?

It is not like mpv will ever have changes that significant that they cannot be ported quickly. See `shared-script-properties`, deprecated in 0.37, removed in 0.38. Highly used feature, scripts were certainly broken by it. But they got updated, at least some of them.

I won't circle back on other points, because everything has been said already. I feel like we are overthinking it, in my opinion having consistent deprecation policy would not only help us avoid thinking about it, but also users to understand when each feature would disappear.

Dudemanguy commented:
>However, when I checked the wiki and manual, I couldn't find any guidance on how to migrate. (In fact, with enough information, I could have known that I should set its sub-option, such as --icc-cache-dir. )

This is a real-time demonstration of what I meant by that deprecated alias being user hostile. It's not your fault. The name is extremely misleading and judging from this response you completely misunderstood what it did.

>Keeping OPT_REMOVED for all removed options is useful to the user and doesn't have any maintenance burden, therefore I support keeping OPT_REMOVED forever for all removed options

I've stated this earlier but I don't agree that it is forever useful and given enough time it would be noise or possibly even confusing if someone happens to type in some ancient obscure option from a decade ago. We've removed plenty of options before with no special message left behind. Apparently I am in the minority on this or something so maybe someone make PR documenting the proposed policy change.

>We have already pretty long period, that certainly gives more than enough time for anyone to adjust to new changes if they want to.
>Even if we were to "break" --slang, two full releases is enough notification period for this.

It's sounds like our divergence is that you think 2 full releases is sufficient for everything. I don't. There's no reason we would ever do this but if we completely break `--slang` I would not consider 2 full releases to be enough. It's not just users having to update their config files, but undoubtedly there's a lot of scripts and other things out there that use this. I don't know what a good number would be but something more like a 5 year deprecation period (so roughly 5 releases if we assume 1 release per year) would be more in the ballpark on that.

>I feel like we are overthinking it, in my opinion having consistent deprecation policy would not only help us avoid thinking about it, but also users to understand when each feature would disappear.

Yeah I think we don't disagree on this part as much as I initially thought. Currently, the only policy we have is basically "important stuff should be deprecated for a while; unimportant stuff can be removed whenever you feel like it". So to bring up the `--cdda-toc-bias` example again, I would consider that option to be obscure and unimportant so we're basically free to break it at anytime. And I've broken obscure options immediately in the past before. But in this case, I ended up waiting the 2 release cycle anyways so it's rather moot.

Anyway, if you want to change it to where *everything* must have at least some minimum deprecation period. I would not be opposed to that. My only pushback is that the difference between "important" and "unimportant" stuff is still meaningful and "important" things would likely need an even *longer* deprecation period than whatever the minimum is. So in my view, that's not technically a uniform policy however changes like that should be extremely rare at this stage. There should virtually be no reason why any of us would need to change any of the well-known highly used options in a non-compatible way.

>We are saying that until it works correctly with OPT_REPLACED there is no reason to actually remove it.

That has never been my interpretation. If we are actually saying this, documentation should be updated.

kasper93 commented:
> This is a real-time demonstration of what I meant by that deprecated alias being user hostile. It's not your fault. The name is extremely misleading and judging from this response you completely misunderstood what it did.

Exactly and that's why `OPT_REPLACED` and `OPT_REMOVED` resolves this ambiguity, by printing the replacement option. What you suggest is removing it completely and leaving user up to themselves to find how the option is called now. @hooke007 got the wrong option, if only there was a way to show the correct one with `OPT_REMOVED`...

As @avih said:
> therefore it only served as a note to users "this doesn't exist anymore, do this or that instead", which is cheap for us and useful to the users.

> 5 year deprecation period (so roughly 5 releases if we assume 1 release per year) would be more in the ballpark on that.

Dude... no one does that. I mean really. There are two kinds of users/script integrators, the ones that are active and will manage to fix it in 2 releases (which is 1-2 years) and the others that are not active or don't care and will fix it only after it becomes hard error, or never.

Side note, removal should be done on the beginning of release period to give most amount of time before stable release for latecomers.

I would consider longer deprecation period only if we break completely libmpv api, requiring significant development effort to integrate new version. Or if we deprecate the mpv custom shader hook in favor of new system. Or deprecate `vo=gpu`. But this kind of deprecation are not relevant for `OPT_REPLACED` and `OPT_REMOVED` that we are discussing here... And likely would stay very long time as dead part of mpv, same as few other vo's currently. 

I agree some deprecations can be left for a long time, but I focused on `OPT_*` part, which are really rarely that hard to port that would require that.

> Anyway, if you want to change it to where everything must have at least some minimum deprecation period.

We have it in the documentation:
https://github.com/mpv-player/mpv/blob/bab9b2cf4ec148f4759f85d62233702f46c23374/DOCS/compatibility.rst?plain=1#L55-L59

I agree, that some things may not be important, but still it doesn't harm us to just wait a bit. Doesn't matter if we remove now or later, if the option is not used anyway.

> That has never been my interpretation. If we are actually saying this, documentation should be updated.

The whole point of `OPT_REPLACED` is to deprecate the option, and notice that it may disappear at any time. But what is the point of removing them, if they still work perfectly fine? And well, the removal should be done with `OPT_REMOVED`.

kasper93 commented on video/out/gpu/video.c:
```diff
@@ -469,12 +466,6 @@ const struct m_sub_options gl_video_conf = {
         {"gpu-shader-cache-dir", OPT_STRING(shader_cache_dir), .flags = M_OPT_FILE},
         {"gpu-hwdec-interop",
             OPT_STRING_VALIDATE(hwdec_interop, ra_hwdec_validate_opt)},
-        {"gamut-warning", OPT_REMOVED("Replaced by --gamut-mapping-mode=warn")},
-        {"gamut-clipping", OPT_REMOVED("Replaced by --gamut-mapping-mode=desaturate")},
-        {"tone-mapping-desaturate", OPT_REMOVED("Replaced by --tone-mapping-mode")},
-        {"tone-mapping-desaturate-exponent", OPT_REMOVED("Replaced by --tone-mapping-mode")},
```
Those are good examples of useful `OPT_REMOVED` usage.

Dudemanguy replied:
I've readded these and left the rest removed.

kasper93 commented on video/out/gpu/video.c:
```diff
@@ -469,12 +466,6 @@ const struct m_sub_options gl_video_conf = {
-        {"gamut-clipping", OPT_REMOVED("Replaced by --gamut-mapping-mode=desaturate")},
-        {"tone-mapping-desaturate", OPT_REMOVED("Replaced by --tone-mapping-mode")},
-        {"tone-mapping-desaturate-exponent", OPT_REMOVED("Replaced by --tone-mapping-mode")},
-        {"tone-mapping-crosstalk", OPT_REMOVED("Hard-coded as 0.04")},
-        {"tone-mapping-mode", OPT_REMOVED("no replacement")},
```
This not that useful, could be just removed.

kasper93 commented on video/out/vulkan/context.c:
```diff
@@ -123,7 +123,6 @@ const struct m_sub_options vulkan_conf = {
         {"vulkan-queue-count", OPT_INT(queue_count), M_RANGE(1, 8)},
         {"vulkan-async-transfer", OPT_BOOL(async_transfer)},
         {"vulkan-async-compute", OPT_BOOL(async_compute)},
-        {"vulkan-disable-events", OPT_REMOVED("Unused")},
```
Not useful.

kasper93 commented on video/out/gpu/video.c:
```diff
@@ -469,12 +466,6 @@ const struct m_sub_options gl_video_conf = {
-        {"gamut-clipping", OPT_REMOVED("Replaced by --gamut-mapping-mode=desaturate")},
-        {"tone-mapping-desaturate", OPT_REMOVED("Replaced by --tone-mapping-mode")},
-        {"tone-mapping-desaturate-exponent", OPT_REMOVED("Replaced by --tone-mapping-mode")},
-        {"tone-mapping-crosstalk", OPT_REMOVED("Hard-coded as 0.04")},
```
This is good, to notify user.

kasper93 commented on video/out/drm_common.c:
```diff
@@ -81,8 +81,6 @@ const struct m_sub_options drm_conf = {
             .help = drm_connector_opt_help},
         {"drm-mode", OPT_STRING_VALIDATE(mode_spec, drm_validate_mode_opt),
             .help = drm_mode_opt_help},
-        {"drm-atomic", OPT_CHOICE(drm_atomic, {"no", 0}, {"auto", 1}),
-            .deprecation_message = "this option is deprecated: DRM Atomic is required"},
```
No user migration needed, can be directly removed. Or `OPT_REMOVED("this option is removed: DRM Atomic is required")`

Dudemanguy commented:
>But I never said that you can't ever remove them. You can remove them with OPT_REMOVED entries.

You just complained that I removed 10 year old deprecated aliases, but OK. Formalize this. In what circumstances is it OK to remove them? 15 years? Only if the thing that it is aliasing also disappears? With regards to OPT_REMOVED, is this required when removing any option or only in certain circumstances? How long should an OPT_REMOVED be in the codebase? 10 years? Forever? If I'm in a retirement home and mpv is still maintained, should it still have an OPT_REMOVED informing users that --gamut-clipping was replaced by --gamut-mapping-mode=desaturate 5 decades ago (assume for the sake of this argument that --gamut-mapping-mode is still an option)? Clearly you feel very strongly about this so formalize this in the documentation and make a PR.

>You only want to apply OPT_REMOVED to arbitrary restricted situations and only for an arbitrary amount of time, because you personally think that they aren't useful to users otherwise. Someone already expressed their views on this right under your post, expressing otherwise.

That user expressed that view based on a complete misunderstanding of an option (explained later). I fail to see how this backs you up in anyway. You used arbitrary as a pejorative here, but the process is fundamentally already arbitrary. Developers make judgement calls on deciding what should be removed, renamed, etc. There is no hard rule for this and can be for a variety of reasons. Your view that OPT_REMOVED should never be removed is just as arbitrary as my view that after enough time passes, it becomes useless code. It is made based upon individual judgements/opinions on what software development practices should be.

>It's not existing in a meaningful way at the current state, and requires continuous effort to "clean up" manually. Using it in the intended way or deleting the system completely would relieve us of this manual labor.

This is what I would consider yet another argument made in bad faith. What you tacitly claim is "the intended way" is nothing more than your personal opinion. It is perfectly fine to think I am an idiot that uses the system in a poor way and so on. But you cannot claim to know what the true intent is. The only person that can do that is the one who actually wrote it which is not any of us in this thread. As for the "manual labor" point, I would file that in a "not an argument" category. mpv is maintained software, so developers will have to crack open some files and change things. It's how it is. If you wish to eliminate the "manual labor" of maintaining options/properties/apis, etc., then the only conclusion is to never remove any deprecations (or maybe never deprecate in the first place but that's quite unfeasible) and leave these things in the codebase forever. If that is the policy, it needs to be documented. Developers will naturally remove old deprecated things to clean up codebases.

>You're just conveniently ignoring the fact that the existing users don't know how to migrate now because you're unwilling to leave an OPT_REMOVED entry. It would've helped the user to migrate.

An `OPT_REMOVED` entry would have been **harmful** in this specific case. I don't wish to drag the user into this bickering but from my understanding what they actually wanted was `--icc-cache-dir`. `--cache-dir` is the **wrong** option. It never did what they wanted. If the user had blindly obeyed a generic "this was replaced by --demuxer-cache-dir" message, they would not have gotten what the desired behavior. In this case, the hard breakage is the *optimum* outcome. It forced the user to consult documentation and realize that the correct option was something else completely. A classic [XY Problem](https://xyproblem.info/) example. In this light, one could even argue that having the alias in the first place is a mistake and I should have completely removed it to avoid this situation. Alternatively, I suppose we could make the `OPT_REMOVED` print out a giant novel detailing and duplicating everything that's already in the manual to avoid this. That seems rather silly to me because consulting the documentation takes a couple of seconds, but OK there's no technical limitation if you insist that is the best approach.

And also I don't know why you keep acting like there was no warning that tells users how to migrate. It was just deemed no longer needed and thus deleted. If you think this guidance should last longer/indefinitely/etc. in some form, please elaborate on what your proposed policy like I said earlier.

>I know what's good and what's not for our users, and if some users don't use mpv in the way we want, well, fuck'em
>fuck everyone who dares to update mpv less frequently than we want them to.

Very nice of you to go into baseless personal attacks and completely misrepresent my arguments.


>There are two kinds of users/script integrators, the ones that are active and will manage to fix it in 2 releases (which is 1-2 years) and the others that are not active or don't care and will fix it only after it becomes hard error, or never.

This is completely speculative, but I think this is optimistic. If we completely break something major that is used all the time, I would expect the pain to last many years. But I suppose there's no way to know without trying, not that we should. I can't think offhand of something that we did in the past that would really fit this category. But maybe you're right and it would all be fixed pretty fast. Who knows.

>We have it in the documentation:

Right but section there clearly says "Important/often used parts". That's not *everything*. Right below, it says "Less useful parts can be broken immediately".

P.S. thanks for the actual PR feedback. I appreciate it. I have no intention of arguing forever with you.

Dudemanguy commented on DOCS/interface-changes/option-removal.txt:
```diff
@@ -0,0 +1,12 @@
+remove deprecated `--cache-dir` option alias
+remove deprecated `--cache-unlink-files` option alias
+remove deprecated `--demuxer-cue-codepage` option alias
+remove deprecated `--fps` option alias
+remove deprecated `--override-display-fps` option alias
+remove deprecated `--sub-ass-force-style` option alias
+remove deprecated `--cdrom-device` option alias
+remove deprecated `--play-dir` option alias
+remove deprecated `--sub-forced-only` option alias
+remove deprecated `--vo-sixel-exit-clear` option alias
+remove deprecated `--cdda-toc-bias` option
+remove deprecated `--drm-atomic` option
```
Are we counting differently? All of these were options were deprecated in 0.37, no? That's 1 release. 0.38 is 2 releases. So the upcoming release, 0.39 is fair game by that rule.

po5 commented:
> Yes and eventually the endgoal is to make them unable to use it altogether.

Punishing the user for using `--play-dir` that was around for 4 years instead of the new `--play-direction` is silly, even if you think the new name is better.
We shouldn't actively try to break user configs.

> But again, we removed literally dozens of options and deprecations in https://github.com/mpv-player/mpv/pull/12436 with virtually no complaints.

You shouldn't expect complaints to come in overnight.
In my case I had a functionality [in my Lua script](https://github.com/po5/trackselect/issues/4) that was silently broken in 5e0902cc94fea62f1641683c9c9fdd5da21d6925, a removal that had no benefits to anyone. It's only after someone noticed and complained that I could investigate why the functionality broke.

Despite what the code comments said, the `track-switched` event was not strictly equivalent to observing `track-list` as it doesn't fire on file load which is useful when you want to detect tracks the user explicitly switched to.
It's worth looking at the intention behind deprecations, where f5c2e3d6e2372546c46dd32ceab1a5c41e19465b tells us it was "for the sake of notifying API users", and that the events aren't meant to be removed (notice the use of the word "never", which I guess Dudemanguy needs to see to be convinced) unless there is a technical reason for it.
This would avoid unnecessary breakage, including silent hard to track down cases like above where the replacement isn't even truly a replacement, and all code for the deprecated functionality needs to be kept for internal use anyway.

Dudemanguy commented:
>Punishing the user for using --play-dir that was around for 4 years instead of the new --play-direction is silly, even if you think the new name is better.

It was actually originally `--play-direction` renamed to `--play-dir` and then I unintentionally renamed it back to the original name. But anyways, I don't feel strongly about that particular one. If you feel it's too soon, we can hold off on it. (surprised nobody has complained about `sub-ass-force-style` yet tbh).

>In my case I had a functionality https://github.com/po5/trackselect/issues/4 that was silently broken in https://github.com/mpv-player/mpv/commit/5e0902cc94fea62f1641683c9c9fdd5da21d6925, a removal that had no benefits to anyone.

I'm sorry that happened, but it's really to comparable to anything here or in the other PR. No 
actual functionality was removed in either one. As for your issue, I don't believe any of the developers were ever aware of it. Why not open up an issue for it? I don't believe there currently is one. We've undeprecated things before.

kasper93 reviewed (approved):
Looks ok, also could be complemented by https://github.com/mpv-player/mpv/pull/14396

Dudemanguy commented:
I left the deprecated aliases for `--play-dir`, `--sub-ass-force-style`, and `--override-display-fps` for now.

kasper93 commented:
@Dudemanguy:

Encoding profiles has to be updated https://github.com/mpv-player/mpv/blob/22fc2e6af3a89db325b3191a3775a373020d562f/etc/encoding-profiles.conf they are using `--vf=fps`. 

```
[cplayer] Option vf-add: 'fps' isn't supported.
[cplayer] Error parsing option vf-add (option parameter could not be parsed)
[cplayer] encoding-profiles.conf:145: setting option vf-add='fps=25' failed.
[cplayer] Option vf-add: 'fps' isn't supported.
[cplayer] Error parsing option vf-add (option parameter could not be parsed)
[cplayer] encoding-profiles.conf:156: setting option vf-add='fps="24000/1001"' failed.
[cplayer] Option vf-add: 'fps' isn't supported.
[cplayer] Error parsing option vf-add (option parameter could not be parsed)
[cplayer] encoding-profiles.conf:175: setting option vf-add='fps=25' failed.
```

Good example of updating things after they break, but let's not go back to that discussion :)

Dudemanguy commented:
Huh? Shouldn't that be the ffmpeg filter?

kasper93 commented:
Maybe I broke something, don't know, nvm.
